File: aoc/2025/aoc202510.hs
module AOC202510 where

import Data.Bits(shiftL,xor,(.|.))
import Data.List(nub)
import Data.Map(Map,adjust,fromList,toList,(!))
import qualified Data.Map
import Data.Set(empty,insert,member)

import AOC

aoc = AOC {
    day="10",
    aocTests=[
        AOCTest {
            testData=unlines [
                "[.##.] (3) (1,3) (2) (2,3) (0,2) (0,1) {3,5,4,7}",
                "[...#.] (0,2,3,4) (2,3) (0,4) (0,1,2) (1,2,3,4) {7,5,12,7,2}",
                "[.###.#] (0,1,2,3,4) (0,3,4) (0,1,2,4,5) (1,2) {10,11,11,5,10,5}"
            ],
            testResult=Just "7",
            testResult2=Just "33"
            }
        ],
    aocCode=ParallelCode {
        pcodeParse=const parse,
        pcodeParse2=const parse2,
        pcodeTest=result,
        pcodeTest2=result2,
        pcodeResult=result,
        pcodeResult2=result2
        }
    }

parse :: String -> [(Int,[Int],[Int])]
parse = map (p . words) . lines
  where
    p (lights:rest) =
        (plights 0 1 lights,map pswitch (init rest),parseInts (last rest))
    plights bits bit ('[':rest) = plights bits bit rest
    plights bits bit (']':rest) = bits
    plights bits bit ('#':rest) = plights (bits .|. bit) (shiftL bit 1) rest
    plights bits bit ('.':rest) = plights bits (shiftL bit 1) rest
    pswitch = sum . map (shiftL 1) . parseInts

result :: Int -> [(Int,[Int],[Int])] -> Int
result ncpu = parallelMapReduce ncpu countPresses sum

countPresses (0,_,_) = 0
countPresses (goal,switches,_) = count 1 switches [] empty
  where
    count npresses [] queue seen = count (npresses+1) queue [] seen
    count npresses (r:rest) queue seen
      | r == goal = npresses
      | member r seen = count npresses rest queue seen
      | otherwise =
          count npresses rest (map (xor r) switches ++ queue) (insert r seen)

parse2 :: String -> [([[Int]],[Int])]
parse2 = map (p . drop 1 . words) . lines
  where p spec = (map parseInts (init spec),parseInts (last spec))

result2 ncpu = parallelMapReduce ncpu countPresses2 sum

countPresses2 :: ([[Int]],[Int]) -> Int
countPresses2 (buttons,joltages)
  | null bsMatrix = bsPresses
  | otherwise = bsPresses + bfPresses
  where
    initialMaxPresses = [minimum [j | (i,j) <- zip [0..] joltages,
                                      i `elem` button]
                         | button <- buttons]
    initialMatrix =
        [[if i `elem` button then 1 else 0 | button <- buttons]
         | (i,joltage) <- zip [0..] joltages]
    initialRHS = joltages

    (udMaxPresses,udMatrix,udRHS) =
        upperDiagonalize initialMaxPresses initialMatrix initialRHS 0

    (bsPresses,bsMaxPresses,bsMatrix,bsRHS) =
        backsubstitute 0 (reverse udMaxPresses)
                         (reverse (map reverse udMatrix))
                         (reverse udRHS)
                         [] []

    bfPresses = bruteForce bsMaxPresses bsMatrix bsRHS

upperDiagonalize :: [Int] -> [[Int]] -> [Int] -> Int -> ([Int],[[Int]],[Int])
upperDiagonalize maxPresses matrix rhs row
  | row >= length matrix = (maxPresses,matrix,rhs)
  | row >= length pivotRow && pivotRHS == 0 = upperDiagonalize maxPresses mat4 rhs4 row
  | row >= length pivotRow = error "bad input"
  | pivot /= 0 = upperDiagonalize maxPresses mat1Div rhs1Div (row+1)
  | not (null pivotRowIndexes2) = upperDiagonalize maxPresses mat2 rhs2 row
  | not (null pivotColIndexes3) = upperDiagonalize maxPresses3 mat3 rhs row
  | head (drop row rhs) == 0 = upperDiagonalize maxPresses mat4 rhs4 row
  | otherwise = error "bad input"
  where
    pivotRow = head $ drop row matrix
    pivot = head $ drop row pivotRow
    pivotRHS = head $ drop row rhs

    mat1 = take (row+1) matrix ++ map udRow (drop (row+1) matrix)
    rhs1 = take (row+1) rhs ++ map udRhs (drop (row+1) (zip matrix rhs))
    udRow lowerRow = [pivot*l - a*p | (l,p) <- zip lowerRow pivotRow]
      where a = head $ drop row lowerRow
    udRhs (lowerRow,lowerRHS) = pivot*lowerRHS - a*pivotRHS
      where a = head $ drop row lowerRow

    (mat1Div,rhs1Div) = divideGCD mat1 rhs1
    divideGCD [] [] = ([],[])
    divideGCD (row:mat) (r:rhs)
      | f == 0 = (row:matDiv,r:rhsDiv)
      | otherwise = (map (`div` f) row:matDiv,r `div` f:rhsDiv)
      where
        (matDiv,rhsDiv) = divideGCD mat rhs
        f = foldr gcd r row

    pivotRowIndexes2 = [i | (i,lowerRow) <- drop (row+1) (zip [0..] matrix),
                            head (drop row lowerRow) /= 0]
    pivotRowIndex2 = head pivotRowIndexes2
    mat2 = take row matrix ++ drop pivotRowIndex2 matrix
                           ++ drop row (take pivotRowIndex2 matrix)
    rhs2 = take row rhs ++ drop pivotRowIndex2 rhs
                        ++ drop row (take pivotRowIndex2 rhs)

    pivotColIndexes3 = [i | (i,coef) <- drop (row+1) (zip [0..] pivotRow),
                            coef /= 0]
    pivotColIndex3 = head pivotColIndexes3
    mat3 = map shiftCols matrix
    maxPresses3 = shiftCols maxPresses
    shiftCols r = take row r ++ drop pivotColIndex3 r
                             ++ drop row (take pivotColIndex3 r)

    mat4 = take row matrix ++ drop (row+1) matrix
    rhs4 = take row rhs ++ drop (row+1) rhs

backsubstitute :: Int -> [Int] -> [[Int]] -> [Int] -> [[Int]] -> [Int] -> (Int,[Int],[[Int]],[Int])
backsubstitute presses maxPresses [] [] topMatrix topRHS =
    (presses,maxPresses,reverse topMatrix,reverse topRHS)
backsubstitute presses maxPresses (row:matrix) (r:rhs) topMatrix topRHS
  | length coefs == 1 = backsubstitute presses1 maxPresses1 mat1 rhs1 [] []
  | otherwise =
      backsubstitute presses maxPresses matrix rhs (row:topMatrix) (r:topRHS)
  where
    coefs = [(i,c) | (i,c) <- zip [0..] row, c /= 0]
    [(col,coef)] = coefs
    sol
      | r `mod` coef /= 0 = error "bad input"
      | r `div` coef < 0 = error "bad input"
      | r `div` coef > head (drop col maxPresses) = error "bad input"
      | otherwise = r `div` coef

    presses1 = presses+sol
    maxPresses1 = rmCol maxPresses
    mat1 = map rmCol (reverse topMatrix) ++ map rmCol matrix
    rhs1 = map subst (reverse (zip topMatrix topRHS)) ++ map subst (zip matrix rhs)

    rmCol rs = take col rs ++ drop (col+1) rs
    subst (rs,rhsVal) = rhsVal - sol*head (drop col rs)

bruteForce :: [Int] -> [[Int]] -> [Int] -> Int
bruteForce maxPresses mat rhs = foldr solve (sum maxPresses) freeVars
  where
    nFreeVars = length maxPresses - length mat
    freeVars = gen (take nFreeVars maxPresses) (take nFreeVars maxPresses)
      where
        gen current maxes
          | all (== 0) current = [current]
          | otherwise = current : gen (next current maxes) maxes
          where
            next [] [] = []
            next (c:cs) (m:ms)
              | c > 0 = c-1 : cs
              | otherwise = m : next cs ms
    solve :: [Int] -> Int -> Int
    solve sols best = subst (sum sols) sols mat rhs
      where
        subst current sols [] [] = min current best
        subst current sols (row:mat) (r:rhs)
          | current >= best = best
          | any (/= 0) (drop (1 + length sols) row) = error "?"
          | sol < 0 = best
          | sol > head (drop (length sols) maxPresses) = best
          | otherwise = subst (current+sol) (sols ++ [sol]) mat rhs
          where
            rhsSubst = r - sum (zipWith (*) row sols)
            coef = head $ drop (length sols) row
            sol
              | rhsSubst `mod` coef /= 0 = best
              | otherwise = rhsSubst `div` coef

File: aoc/2025/aoc202510.go
package main

import (
	"math/bits"
	"runtime"
)

func init() {
	Register(&aoc202510{
		AOC: AOC{
			Day: 10,
			Tests: []Test{
				Test{
					Input: `[.##.] (3) (1,3) (2) (2,3) (0,2) (0,1) {3,5,4,7}
[...#.] (0,2,3,4) (2,3) (0,4) (0,1,2) (1,2,3,4) {7,5,12,7,2}
[.###.#] (0,1,2,3,4) (0,3,4) (0,1,2,4,5) (1,2) {10,11,11,5,10,5}
`,
					Part1: "7",
					Part2: "33",
				},
			},
		},
	})
}

type aoc202510 struct {
	AOC
}

type aoc202510_machine struct {
	lights   int
	buttons  []int
	joltages []int

	buttons2 [][]int
}

func (m aoc202510_machine) configure() int {
	minPresses := len(m.buttons)
	for i := range 1 << len(m.buttons) {
		npresses := bits.OnesCount(uint(i))
		if npresses >= minPresses {
			continue
		}
		lights := 0
		for j, b := range m.buttons {
			if (1<<j)&i != 0 {
				lights ^= b
			}
		}
		if lights == m.lights {
			minPresses = npresses
		}
	}
	return minPresses
}

func (aoc *aoc202510) parse(input *Input) Seq[aoc202510_machine] {
	return func(yield func(aoc202510_machine) bool) {
		for line, ok := input.Line(); ok; line, ok = input.Line() {
			in := InputString(line)
			in.Char()
			lights := 0
			bit := 1
			for ch, _ := in.Char(); ch != ']'; ch, _ = in.Char() {
				if ch == '#' {
					lights |= bit
				}
				bit <<= 1
			}
			var buttons []int
			var buttons2 [][]int
			for {
				w, _ := in.Word()
				if w[0] == '{' {
					if !yield(aoc202510_machine{
						lights:   lights,
						buttons:  buttons,
						joltages: InputString(w).Ints(),
						buttons2: buttons2,
					}) {
						return
					}
					break
				}
				button := 0
				button2 := []int{}
				for _, i := range InputString(w).Ints() {
					button |= 1 << i
					button2 = append(button2, i)
				}
				buttons = append(buttons, button)
				buttons2 = append(buttons2, button2)
			}
		}
	}
}

func (aoc *aoc202510) Part1(input *Input) string {
	in := make(chan aoc202510_machine)
	out := make(chan int)
	for range runtime.NumCPU() {
		go func() {
			result := 0
			for machine := range in {
				result += machine.configure()
			}
			out <- result
		}()
	}
	for machine := range aoc.parse(input) {
		in <- machine
	}
	close(in)

	result := 0
	for range runtime.NumCPU() {
		result += <-out
	}
	return IntResult(result)
}

func (m aoc202510_machine) configure2() int {
	matrix := make([][]int, len(m.joltages))
	maxPresses := make([]int, len(m.buttons))
	for i, rhs := range m.joltages {
		row := make([]int, len(m.buttons)+1)
		row[len(m.buttons)] = rhs
		matrix[i] = row
		for j, button := range m.buttons {
			if button&(1<<i) != 0 {
				maxPresses[j] = max(maxPresses[j], rhs)
				row[j] = 1
			}
		}
	}

	// to upper diagonal
	for i := range matrix {
		if i >= len(matrix[i])-1 {
			break
		}
		if matrix[i][i] != 0 {
			goto foundPivot
		}
		for j := i + 1; j < len(matrix); j++ {
			if matrix[j][i] != 0 {
				matrix[i], matrix[j] = matrix[j], matrix[i]
				goto foundPivot
			}
		}
		for j := i + 1; j < len(matrix[i])-1; j++ {
			if matrix[i][j] != 0 {
				maxPresses[i], maxPresses[j] = maxPresses[j], maxPresses[i]
				for k := range matrix {
					matrix[k][i], matrix[k][j] = matrix[k][j], matrix[k][i]
				}
				goto foundPivot
			}
		}
		continue
	foundPivot:
		for j := i + 1; j < len(matrix); j++ {
			ai := matrix[i][i]
			aj := matrix[j][i]
			for k := i; k < len(matrix[j]); k++ {
				matrix[j][k] = matrix[j][k]*ai - matrix[i][k]*aj
			}
		}
	}

	presses := 0
	// backsubstitute
	substitutionMade := true
	for substitutionMade {
		substitutionMade = false
		for i := len(matrix) - 1; i >= 0; i-- {
			column := -1
			for j := range len(matrix[i]) - 1 {
				if matrix[i][j] != 0 {
					if column == -1 {
						column = j
					} else {
						column = -2
						break
					}
				}
			}
			if column == -2 {
				continue
			} else if column == -1 {
				copy(matrix[i:], matrix[i+1:])
				matrix = matrix[:len(matrix)-1]
				continue
			}
			substitutionMade = true
			p := matrix[i][len(matrix[i])-1] / matrix[i][column]
			if p*matrix[i][column] != matrix[i][len(matrix[i])-1] {
				panic("bad input")
			}
			if p < 0 {
				panic("bad input")
			}
			if p > maxPresses[column] {
				panic("bad input")
			}
			presses += p
			copy(maxPresses[column:], maxPresses[column+1:])
			maxPresses = maxPresses[:len(maxPresses)-1]
			for i, row := range matrix {
				row[len(row)-1] -= row[column] * p
				copy(row[column:], row[column+1:])
				matrix[i] = row[:len(row)-1]
			}
		}
	}

	neqs := len(matrix)
	if neqs == 0 {
		return presses
	}
	nvars := len(matrix[0]) - 1
	best := 0
	for _, maxPress := range maxPresses {
		best += maxPress
	}
	sol := make([]int, nvars)
	sol[neqs] = -1
bruteForceLoop:
	for {
		more := false
		for i := neqs; i < nvars; i++ {
			sol[i]++
			if sol[i] <= maxPresses[i] {
				more = true
				break
			}
			sol[i] = 0
		}
		if !more {
			return presses + best
		}
		for i := neqs - 1; i >= 0; i-- {
			rhs := matrix[i][nvars]
			for j := i + 1; j < nvars; j++ {
				rhs -= matrix[i][j] * sol[j]
			}
			sol[i] = rhs / matrix[i][i]
			if sol[i] < 0 || sol[i]*matrix[i][i] != rhs {
				continue bruteForceLoop
			}
		}
		sum := 0
		for _, s := range sol {
			sum += s
		}
		best = min(best, sum)
	}
}

func (aoc *aoc202510) Part2(input *Input) string {
	in := make(chan aoc202510_machine)
	out := make(chan int)
	for range runtime.NumCPU() {
		go func() {
			result := 0
			for machine := range in {
				result += machine.configure2()
			}
			out <- result
		}()
	}

	for machine := range aoc.parse(input) {
		in <- machine
	}
	close(in)

	result := 0
	for range runtime.NumCPU() {
		result += <-out
	}
	return IntResult(result)
}
